### PR TITLE
Initial commit for F# test runner

### PR DESCRIPTION
**Pull request by umaranis**

### CODE REVIEW

**Comment by ErikSchierboom** [comment]
Thanks! I'll do a review next week due to me still being on holiday.

File: run.ps1 [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+function Run-All-Tests {
+    Remove-Item $resultsTrxFile -ErrorAction Ignore
+
+    $logger = "trx;LogFileName=results.trx"
+    dotnet test $InputDirectory --logger:$logger | Out-String | Sanitize-Test-Output
+}
+
+function Get-Tests-Ordered-By-Line-Number {
+    $testFileContents = Get-Content $testFile
+    $testMethodRegex = "let\s+````([\w ]+)````\s+\("
```
Comment (ErikSchierboom) [inline run.ps1]: I'm not entirely sure it works, but you may try something like this (a greedy match):

```suggestion
    $testMethodRegex = "let\s+````(.+?)````\s+\("
```

This version will also support using things like apostrophes and such.

File: tests/Exercism.TestRunner.FSharp.IntegrationTests/DirectoryHelper.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
+using System;
+using System.IO;
+using System.Linq;
+
+namespace Exercism.TestRunner.FSharp.IntegrationTests
```
Comment (ErikSchierboom) [inline tests/Exercism.TestRunner.FSharp.IntegrationTests/DirectoryHelper.cs]: Great to have this working. I was wondering if you'd be interested in converting all the C# code to F#? If not, we can merge this PR as is and convert to F# later.

File: tests/Exercism.TestRunner.FSharp.IntegrationTests/Solutions/CompileError/Fake.cs [hunk 3]
```diff
@@ -0,0 +1,6 @@
+using Xunit;
```
Comment (ErikSchierboom) [inline tests/Exercism.TestRunner.FSharp.IntegrationTests/Solutions/CompileError/Fake.cs]: This test file should be an F# file, as the integration tests should run against F# projects.

File: tests/Exercism.TestRunner.FSharp.IntegrationTests/Solutions/CompileError/Fake.csproj [hunk 4]
```diff
@@ -0,0 +1,17 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (ErikSchierboom) [inline tests/Exercism.TestRunner.FSharp.IntegrationTests/Solutions/CompileError/Fake.csproj]: This project file should be an F# project files (.fsproj), as the integration tests should run against F# projects. For a sample F# project file, see [this file](https://github.com/exercism/fsharp/blob/master/exercises/accumulate/AccumulateTest.fs).

File: tests/Exercism.TestRunner.FSharp.IntegrationTests/Solutions/CompileError/Fake.cs (hunk 3 above)
Comment (ErikSchierboom) [reply tests/Exercism.TestRunner.FSharp.IntegrationTests/Solutions/CompileError/Fake.cs]: Note that this comment applies to all the C# files within the `tests/Exercism.TestRunner.FSharp.IntegrationTests/Solutions` directory.

File: tests/Exercism.TestRunner.FSharp.IntegrationTests/Solutions/CompileError/Fake.csproj (hunk 4 above)
Comment (ErikSchierboom) [reply tests/Exercism.TestRunner.FSharp.IntegrationTests/Solutions/CompileError/Fake.csproj]: Note that this comment applies to all the F# project files within the `tests/Exercism.TestRunner.FSharp.IntegrationTests/Solutions` directory.

File: .editorconfig [hunk 5]
```diff
@@ -0,0 +1,147 @@
+###############################
```
[the remark this answers is not in the document]
Comment (umaranis) [reply .editorconfig]: Even though we are running F# tests, the runner code is still in C#. I guess we still need this file till we port all code over to F#.

File: format.ps1 [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+    Formats the .NET source code, as well as all markdown and JSON files.
+.EXAMPLE
+    The example below will format all source code
+    PS C:\> ./format.ps1
+.NOTES
+    The formatting of markdown and JSON files is done through prettier. This means
+    that NPM has to be installed for this functionality to work.
+#>
+
+dotnet tool install -g dotnet-format --version 3.0.4
```
[the remark this answers is not in the document]
Comment (umaranis) [reply format.ps1]: Again, should we keep it till all the code is ported over to F#?

File: .editorconfig (hunk 5 above)
[the remark this answers is not in the document]
Comment (ErikSchierboom) [reply .editorconfig]: Correct. So this comment only applies when we port the code over.

File: format.ps1 (hunk 6 above)
[the remark this answers is not in the document]
Comment (ErikSchierboom) [reply format.ps1]: Correct. So this comment only applies when we port the code over.

File: tests/Exercism.TestRunner.FSharp.IntegrationTests/DirectoryHelper.cs (hunk 2 above)
Comment (umaranis) [reply tests/Exercism.TestRunner.FSharp.IntegrationTests/DirectoryHelper.cs]: I would suggest merging this PR. 

I will convert the code to F# later.

File: tests/Exercism.TestRunner.FSharp.IntegrationTests/Solutions/CompileError/Fake.cs (hunk 3 above)
Comment (umaranis) [reply tests/Exercism.TestRunner.FSharp.IntegrationTests/Solutions/CompileError/Fake.cs]: Ported all projects `tests/Exercism.TestRunner.FSharp.IntegrationTests/Solutions` directory to F#

File: tests/Exercism.TestRunner.FSharp.IntegrationTests/Solutions/CompileError/Fake.csproj (hunk 4 above)
Comment (umaranis) [reply tests/Exercism.TestRunner.FSharp.IntegrationTests/Solutions/CompileError/Fake.csproj]: Ported all projects `tests/Exercism.TestRunner.FSharp.IntegrationTests/Solutions` directory to F#

File: tests/Exercism.TestRunner.FSharp.IntegrationTests/DirectoryHelper.cs (hunk 2 above)
Comment (ErikSchierboom) [reply tests/Exercism.TestRunner.FSharp.IntegrationTests/DirectoryHelper.cs]: Sure, will do!

**Comment by ErikSchierboom** [comment]
@kytrinyx Could you perhaps enable Travis builds for this repository?

File: tests/Exercism.TestRunner.FSharp.IntegrationTests/DirectoryHelper.cs (hunk 2 above)
Comment (ErikSchierboom) [reply tests/Exercism.TestRunner.FSharp.IntegrationTests/DirectoryHelper.cs]: Works for me!

File: tests/Exercism.TestRunner.FSharp.IntegrationTests/Solutions/CompileError/expected_results.json [hunk 7]
```diff
@@ -0,0 +1,5 @@
+{
+  "status": "error",
+  "message": "FakeTest.fs(9,5): error FS0039: The value or constructor 'Add' is not defined.",
```
[the remark this answers is not in the document]
Comment (umaranis) [reply tests/Exercism.TestRunner.FSharp.IntegrationTests/Solutions/CompileError/expected_results.json]: This one was failing for me too. I wanted to ask you what should the path be? 

PS: The error message contains the full path which is user-dependent, so I can't include that. Maybe we need to get it from Travis after integration.

[the remark this answers is not in the document]
Comment (ErikSchierboom) [reply tests/Exercism.TestRunner.FSharp.IntegrationTests/Solutions/CompileError/expected_results.json]: What I would like is for the path to be stripped up until the filename. `
/Users/erikschierboom/Programming/exercism/fsharp-test-runner/tests/Exercism.TestRunner.FSharp.IntegrationTests/Solutions/CompileError/FakeTest.fs` should thus become `FakeTest.fs`. There already is some sanitizing code in the script: https://github.com/exercism/fsharp-test-runner/pull/2/files#diff-aa9c79ea204eadab08626b7af98248dcR50 Maybe you could modify that or add to it to make it work?

[the remark this answers is not in the document]
Comment (ErikSchierboom) [reply tests/Exercism.TestRunner.FSharp.IntegrationTests/Solutions/CompileError/expected_results.json]: The reason for stripping the path is that this path is irrelevant to the user, as well as the fact that we don't want to show the path on the build server or test server to the user :)

[the remark this answers is not in the document]
Comment (umaranis) [reply tests/Exercism.TestRunner.FSharp.IntegrationTests/Solutions/CompileError/expected_results.json]: Fixed the issue by removing the path, just leaving the filename.

File: tests/Exercism.TestRunner.FSharp.IntegrationTests/Solutions/MultipleTestsWithMultipleFails/Fake.fs [hunk 8]
```diff
@@ -0,0 +1,9 @@
+module Fake
+
+open Xunit
+
+let Add x y = x + y
```
[the remark this answers is not in the document]
Comment (umaranis) [reply tests/Exercism.TestRunner.FSharp.IntegrationTests/Solutions/MultipleTestsWithMultipleFails/Fake.fs]: Yeah, I agree. 
I wasn't sure of any dependency on the method names mentioned in `expected-results.json`, so tried to keep the changes to the minimum.

I will address this in a future PR.

[the remark this answers is not in the document]
Comment (ErikSchierboom) [reply tests/Exercism.TestRunner.FSharp.IntegrationTests/Solutions/MultipleTestsWithMultipleFails/Fake.fs]: Great, thanks!

File: tests/Exercism.TestRunner.FSharp.IntegrationTests/Solutions/CompileError/expected_results.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (ErikSchierboom) [reply tests/Exercism.TestRunner.FSharp.IntegrationTests/Solutions/CompileError/expected_results.json]: Very nice!

File: tests/Exercism.TestRunner.FSharp.IntegrationTests/Solutions/MultipleTestsWithMultipleFails/expected_results.json [hunk 9]
```diff
@@ -0,0 +1,19 @@
+{
+  "status": "fail",
+  "tests": [
+    {
+      "name": "FakeTest.Add_should_add_numbers",
+      "status": "fail",
+      "message": "FsUnit.Xunit+MatchException : Exception of type 'FsUnit.Xunit+MatchException' was thrown.\nExpected: Equals 3\nActual:   was 2"
```
Comment (ErikSchierboom) [inline tests/Exercism.TestRunner.FSharp.IntegrationTests/Solutions/MultipleTestsWithMultipleFails/expected_results.json]: Maybe we could strip the "FsUnit.Xunit+MatchException : Exception of type 'FsUnit.Xunit+MatchException' was thrown." part from the message too? Sorry that I just noticed this. It's also fine if you'd like to do this in a separate PR.

Comment (umaranis) [reply tests/Exercism.TestRunner.FSharp.IntegrationTests/Solutions/MultipleTestsWithMultipleFails/expected_results.json]: Done. See commit 278b27365a55420d00c3afd2793e8045c4cae97a

Comment (ErikSchierboom) [reply tests/Exercism.TestRunner.FSharp.IntegrationTests/Solutions/MultipleTestsWithMultipleFails/expected_results.json]: Superb!

**Comment by ErikSchierboom** [comment]
Merged. Great work @umaranis! 🎉 

Looking forward to the things coming in future PR's :)
